src/azul/portal_service.py

```diff
@@ -0,0 +1,82 @@
+from copy import deepcopy
+from typing import (
+    Sequence,
+)
+import json
+import logging
+
+import boto3
+from botocore.exceptions import ClientError
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSON
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def get_portal_integrations_db(self) -> Sequence[JSON]:
+        """
+        Retrieve portal DB from S3.
+
+        If no object is present at the location specified in azul.config, then
+        a default hardcoded DB is obtained from the current plugin, uploaded to
+        S3, and returned.
+        """
+
+        client = boto3.client('s3')
+        try:
+            retrieval = client.get_object(Bucket=config.portal_integrations_db_bucket,
+                                          Key=config.portal_integrations_db_object)
+        except client.exceptions.NoSuchKey:
+            plugin = Plugin.load()
+            db = self.stage_db(plugin.portal_integrations_db())
+            log.info('Portal integration DB not found in S3; using and uploading hardcoded DB from plugin.')
+            json_bytes = bytes(json.dumps(db), 'UTF-8')
+            client.put_object(Bucket=config.portal_integrations_db_bucket,
+                              Key=config.portal_integrations_db_object,
+                              Body=json_bytes,
+                              ContentType='application/json')
+        except ClientError as e:
```

Comment:
Also, we prefer using the `exc_info` argument to log methods. You should try it to see what the difference is.

Reply:
I added this to assist in debugging the service response unit tests, which at one point were failing with no diagnostic information besides a 500 internal server error. But I can see how this usefulness doesn't translate into deployment.

Reply:
> which at one point were failing with no diagnostic information besides a 500 internal server error

Can you reproduce this? It may indicate that something is wrong with test logging.

Reply:
Steps to reproduce:

1. Checkout `develop` or `issues/noah-aviel-dove/1383-move-portal-integration-db`
2. Add the line `raise ValueError("Can't see me!")` at the top of the `_fetch_integrations` function in app.py.
3. Right-click on test/service/test_response.py and run unit tests
4. Expected output for the first subtest (entity_type='file', integration_type='get_manifest'):

```
SubTest error: Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 523, in subTest
    yield
  File "/home/nadove/PycharmProjects/azul/test/service/test_response.py", line 1554, in test_integrations
    response_json = self._get_integrations(params)
  File "/home/nadove/PycharmProjects/azul/test/service/test_response.py", line 1518, in _get_integrations
    response.raise_for_status()
  File "/home/nadove/PycharmProjects/azul/.venv/lib/python3.6/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://127.0.0.1:43109/integrations?integration_type=get_manifest&entity_type=file
```

5. To programatically demonstrate that the underlying exception does not appear anywhere in the output:
6. `cd azul/test/service`
7. `python -m unittest test_response.py 2>&1 | fgrep ValueError`
8. Expected output: none.

Reply:
I think I know what's going on: https://github.com/DataBiosphere/azul/issues/1414

That actually contradicts what I said about Chalice handling and logging errors: it only does that when self.debug is set and that's currently tied to AZUL_DEBUG > 0.

As a workaround, when I set AZUL_DEBUG to 1.

Reply:
I'll see if I can quickly solve #1414.